twill-core/src/main/java/org/apache/twill/internal/json/TwillRuntimeSpecificationCodec.java

```diff
@@ -0,0 +1,90 @@
+ */
+package org.apache.twill.internal.json;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import org.apache.twill.api.TwillSpecification;
+import org.apache.twill.internal.TwillRuntimeSpecification;
+
+import java.lang.reflect.Type;
+
+/**
+ * Codec for serializing and deserializing a {@link TwillRuntimeSpecification} object using json.
+ */
+final class TwillRuntimeSpecificationCodec implements JsonSerializer<TwillRuntimeSpecification>,
+                                                         JsonDeserializer<TwillRuntimeSpecification> {
+
+  private static final String FS_USER = "fsUser";
+  private static final String TWILL_APP_DIR = "twillAppDir";
+  private static final String ZK_CONNECT_STR = "zkConnectStr";
+  private static final String TWILL_RUNID = "twillRunId";
+  private static final String TWILL_APP_NAME = "twillAppName";
+  private static final String RESERVED_MEMORY = "reservedMemory";
+  private static final String RM_SCHEDULER_ADDR = "rmSchedulerAddr";
+  private static final String LOG_LEVEL = "logLevel";
+  private static final String TWILL_SPEC = "twillSpecification";
+
+  @Override
+  public JsonElement serialize(TwillRuntimeSpecification src, Type typeOfSrc, JsonSerializationContext context) {
+    JsonObject json = new JsonObject();
+    json.addProperty(TWILL_APP_DIR, src.getTwillAppDir());
+    json.addProperty(ZK_CONNECT_STR, src.getZkConnectStr());
+    json.addProperty(TWILL_RUNID, src.getTwillRunId());
+    json.addProperty(TWILL_APP_NAME, src.getTwillAppName());
+    json.addProperty(RESERVED_MEMORY, src.getReservedMemory());
+    if (src.getFsUser() != null) {
+      json.addProperty(FS_USER, src.getFsUser());
+    }
+    if (src.getRmSchedulerAddr() != null) {
+      json.addProperty(RM_SCHEDULER_ADDR, src.getRmSchedulerAddr());
+    }
+    if (src.getLogLevel() != null) {
+      json.addProperty(LOG_LEVEL, src.getLogLevel());
+    }
+    json.add(TWILL_SPEC, context.serialize(src.getTwillSpecification(),
+                                           new TypeToken<TwillSpecification>() { }.getType()));
+    return json;
+  }
+
+  @Override
+  public TwillRuntimeSpecification deserialize(JsonElement json, Type typeOfT,
+                                                                    JsonDeserializationContext context) throws JsonParseException {
+    JsonObject jsonObj = json.getAsJsonObject();
+
+    TwillSpecification twillSpecification = context.deserialize(
+      jsonObj.get(TWILL_SPEC), new TypeToken<TwillSpecification>() { }.getType());
+    return new TwillRuntimeSpecification(twillSpecification,
+                                         jsonObj.has(FS_USER) ? jsonObj.get(FS_USER).getAsString() : null,
```

Comment:
I don't see constructor of `TwillRuntimeSpecification` has `@Nullable` for the parameters. Please annotate accordingly for parameters that are allowed to have `null` values.